File: OJ practices/LeetCode题目集/106. Construct Binary Tree from Inorder and Postorder Traversal(mdium).java
/*
 * @Author: SourDumplings
 * @Date: 2020-09-05 22:00:05
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
 */

class Solution
{
    public TreeNode buildTree(int[] inorder, int[] postOrder)
    {
        if (inorder.length == 0)
        {
            return null;
        }
        int rootVal = postOrder[postOrder.length - 1];
        int leftEnd = inorder.length;
        for (int i = 0; i < inorder.length; i++)
        {
            if (inorder[i] == rootVal)
            {
                leftEnd = i;
                break;
            }
        }
        TreeNode root = new TreeNode(rootVal);
        root.left = build(inorder, 0, leftEnd, postOrder, 0, leftEnd);
        root.right = build(inorder, leftEnd + 1, inorder.length, postOrder, leftEnd, postOrder.length - 1);
        return root;
    }

    private TreeNode build(int[] inOrder, int inB, int inE, int[] postOrder, int postB, int postE)
    {
        if (inE <= inB)
        {
            return null;
        }
        int rootVal = postOrder[postE - 1];
        TreeNode root = new TreeNode(rootVal);
        int leftEnd = inE;
        for (int i = inB; i < inE; i++)
        {
            if (inOrder[i] == rootVal)
            {
                leftEnd = i;
                break;
            }
        }
        root.left = build(inOrder, inB, leftEnd, postOrder, postB, leftEnd - inB + postB);
        root.right = build(inOrder, leftEnd + 1, inE, postOrder, leftEnd - inB + postB, postE - 1);
        return root;
    }
}

File: OJ practices/LeetCode题目集/106. Construct Binary Tree from Inorder and Postorder Traversal(mdium).cpp
/**
 * @file 106. Construct Binary Tree from Inorder and Postorder Traversal(mdium).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief
 * https://leetcode.cn/problems/construct-binary-tree-from-inorder-and-postorder-traversal/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-12-09
 *
 * @copyright Copyright (c) 2024 SourDumplings
 *
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
	TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder)
	{
		size_t n = inorder.size();
		return buildTree(inorder, 0, n, postorder, 0, n);
	}

private:
	TreeNode *buildTree(
		const vector<int> &in,
		size_t inB,
		size_t inE,
		const vector<int> &post,
		size_t postB,
		size_t postE
	)
	{
		if (inB == inE) { return nullptr; }
		int rootVal	   = post[postE - 1];
		TreeNode *root = new TreeNode(rootVal);
		for (size_t i = inB; i != inE; ++i)
		{
			if (in[i] == rootVal)
			{
				root->left	= buildTree(in, inB, i, post, postB, postB + i - inB);
				root->right = buildTree(in, i + 1, inE, post, postB + i - inB, postE - 1);
			}
		}
		return root;
	}
};